restli-common/src/main/java/com/linkedin/restli/common/validation/RestLiDataValidator.java

```diff
@@ -534,7 +540,9 @@ private ValidationResult checkNewRecordsAreNotMissingFields(RecordTemplate entit
         // Replace $set with the field name to get the full path
         path[path.length - 1] = message.getFormat();
         DataElement element = DataElementUtil.element(new SimpleDataElement(entity.data(), entity.schema()), path);
-        ValidationResult result = ValidateDataAgainstSchema.validate(element, new ValidationOptions());
+        ValidationOptions validationOptions = new ValidationOptions();
+        validationOptions.setTreatOptional(_readOnlyOptionalPredicate);
```

Comment:
This should be done only for $set on array fields, see my comment above.

Reply:
I understand why this would apply to arrays, but I don't think this applies to just arrays. If a field contains an inner record, then setting that field is basically the same as updating the entire record and all its descendants as new records. This logic applies to both arrays and non-arrays. Otherwise, setting a field as an entire record is impossible if that record contains ReadOnly fields.

Also, the user who originally reported this bug is experiencing it for a union-of-a-record type, not an array.

Reply:
Treating them as optional is fine as they won't be present. 

What I meant is that we should also check that ReadOnly fields are **not present** for non array types. ie, Run the readOnlyRestricted predicate. (we cannot do that for arrays due to the reason in my above comment).